src/OrchardCore/OrchardCore.DisplayManagement/Descriptors/FeatureShapeDescriptor.cs

```diff
@@ -44,15 +44,15 @@ public class ShapeDescriptorIndex : ShapeDescriptor
             // pre-calculate as much as we can
             _alternationDescriptors = alterationKeys
                 .Select(key => _descriptors[key])
-                .ToList();
+                .ToArray();
 
             _wrappers = _alternationDescriptors
                 .SelectMany(sd => sd.Wrappers)
```

Comment:
Can perhaps all of these `SelectMany`s optimized away too? So, have a single iteration instead of 8.

Reply:
@Piedone done.